contracts/ligo/src/minter_collection/editions/fa2_multi_nft_token_editions.mligo

```diff
@@ -0,0 +1,121 @@
+#include "../../../fa2_modules/pauseable_admin_option.mligo"
+
+type mint_editions = 
+[@layout:comb]
+{
+    token_metadata : token_metadata;
+    number_of_editions : nat;
+}
+
+type distribute_edition = 
+[@layout:comb] 
+{
+    edition_id : nat; 
+    receivers : address list;
+}
+
+type editions_entrypoints = 
+ | FA2 of nft_asset_entrypoints
+ | Mint_editions of mint_editions list
+ | Distribute_editions of distribute_edition list
+
+type edition_metadata =
+  [@layout:comb]
+  {   
+      creator : address;
+      token_metadata : token_metadata;
+      initial_token_id : nat;
+      number_of_editions : nat;
+      number_of_editions_to_distribute : nat;
+  }
+
+type editions_metadata = (nat, edition_metadata) big_map
+
+type editions_storage = {
+    nft_asset_storage : nft_asset_storage;
+    editions_metadata : editions_metadata;
+    current_edition_id : nat; 
+}
+
+let rec mint_editions ( editions_list , storage : mint_editions list * editions_storage)
+  : operation list * editions_storage =
+  let mint_single_edition_set : (mint_editions * editions_storage) -> editions_storage = fun (param, storage : mint_editions * editions_storage) ->  
+    let edition_metadata : edition_metadata = {
+      creator = Tezos.sender;
+      token_metadata = param.token_metadata;
+      initial_token_id = storage.nft_asset_storage.assets.next_token_id;
+      number_of_editions = param.number_of_editions;
+      number_of_editions_to_distribute = param.number_of_editions 
+    } in
+    let new_editions_metadata = Big_map.add storage.current_edition_id edition_metadata storage.editions_metadata in
+    let new_asset_storage = {storage.nft_asset_storage.assets with next_token_id = storage.nft_asset_storage.assets.next_token_id + param.number_of_editions} in 
+    {storage with 
+        editions_metadata = new_editions_metadata; 
+        nft_asset_storage = {storage.nft_asset_storage with assets = new_asset_storage}
+    } in
+  (match List.head_opt editions_list with 
+    | Some editions -> let new_storage = mint_single_edition_set(editions, storage) in 
+        (match List.tail_opt editions_list with
+          | Some remaining_editions -> mint_editions (remaining_editions, new_storage)
+          | None -> (failwith "INTERNAL_ERROR" : operation list * editions_storage))
+    | None -> ([] : operation list), storage)
+
+let distribute_edition_to_address (edition_metadata, to_, storage : edition_metadata * address * editions_storage) 
+  : editions_storage * edition_metadata = 
+  let token_id = if (edition_metadata.number_of_editions_to_distribute <= 0n) then 
+       (failwith "NO_EDITIONS_TO_DISTRIBUTE" : nat) else 
+       (edition_metadata.initial_token_id + abs (edition_metadata.number_of_editions - edition_metadata.number_of_editions_to_distribute)) in 
+  let new_edition_metadata = {edition_metadata with number_of_editions_to_distribute = abs(edition_metadata.number_of_editions_to_distribute - 1)} in
```

Comment:
this should only be updated once per call to `distribute_edition_to_addresses`, i.e. the loop could look like:
1. get available number of editions, start ID, and creator address
2. for each element in the list, distribute to address
3. update the number of available editions or fail at the end if there are too few for how many list elements we have

Reply:
This also allows us to use an integer counter and not worry about nat addition with negative checks until the end (i.e. don't fail with "out of edition copies to distribute" until the list loop is done)

Reply:
Thanks I just made this change